common/lib/xmodule/xmodule/library_content_module.py

```diff
@@ -371,13 +388,186 @@ def get_child_descriptors(self):
         return list(self._get_selected_child_blocks())
 
 
+@XBlock.wants('user')
 @XBlock.wants('library_tools')  # Only needed in studio
-class AdaptiveLibraryContentModule(AdaptiveLibraryContentFields, LibraryContentModule):
+class AdaptiveLibraryContentModule(AdaptiveLibraryContentFields, LibraryContentModule, AdaptiveLearningAPIMixin):
     """
     A specialized version of Randomized Content Block that communicates
     with an external service to determine when to show its children, and which ones.
     """
 
+    @lazy
+    def parent_course(self):
+        """
+        Return parent course.
+        """
+        return self.runtime.modulestore.get_course(self.course_id, depth=1)
+
+    @lazy
+    def parent_unit_id(self):
+        """
+        Return block ID of parent unit.
+        """
+        parent_unit = self.get_parent()
+        return parent_unit.scope_ids.usage_id.block_id
+
+    @lazy
+    def anonymous_user_id(self):
+        """
+        Return anonymous ID for current user.
+
+        Note that we can't use `user_service.get_anonymous_user_id` here
+        because it only produces meaningful results for staff users.
```

Comment:
Sorry I wasn't clear, @itsjeyd !

If you think upstream would accept it, I'd recommend changing [`xblock_django.user_service.get_anonymous_user_id`](https://github.com/edx/edx-platform/blob/master/common/djangoapps/xblock_django/user_service.py#L34-L55) to something like this:

```python
    def get_anonymous_user_id(self, username, course_id):
        """
        Get the anonymous user id for a user.
        Args:
            username(str): username of a user.
            course_id(str): course id of particular course.
        Returns:
            A unique anonymous_user_id for (user, course) pair.
            None for Non-staff users, unless the current user matches the requested username.
        """
        current_user = self.get_current_user()
        try:
            user = get_user_by_username_or_email(username_or_email=username)
            if not (current_user.opt_attrs.get(ATTR_KEY_USER_ID) == user.id or 
                      current_user.opt_attrs.get(ATTR_KEY_USER_IS_STAFF):
                return None
        except User.DoesNotExist:
            return None

        course_id = CourseKey.from_string(course_id)
        return anonymous_id_for_user(user=user, course_id=course_id, save=False)
```

Then you could use `user_service.get_anonymous_user_id()` directly like you had before in https://github.com/open-craft/edx-platform/pull/49/commits/4904bf47a72900b09159d552cf3c46ebaf0abe39#diff-5b223f1aa12d0cba135cf45dade27de3R500.

But it's difficult to know what the best course of action is.  It would be a major pain if you had to change the anonymous ID algorithm later to get these changes accepted by upstream.